### DISCUSSION TITLE
Unable to change default branch

### DISCUSSION THREAD

**Discussion by gdude2002** [community/community#24324] [Apps, API and Webhooks]
<p>This is an odd one - I don’t seem to be able to change the default branch from <code>master</code> to <code>root</code>, which is what we’d like to use for our repos given the move away from <code>master</code> as the default branch name.</p>
<p>I found <a href="https://github.community/t/cannot-change-default-branch/13728">this topic from 2019</a>, but this doesn’t seem to work even with the correct API calls. GH’s direct contact support is busy and I haven’t gotten a reply yet, so I was wondering if anyone else has run into this. Here’s what happens:</p>
<p><strong>Via the API (using the <code>hub</code> CLI tool)</strong></p>
<ol>
<li><code>hub api repos/&lt;org&gt;/&lt;repo&gt; -X PATCH -F name="&lt;repo&gt;" -F default_branch="root"</code></li>
<li>Note that the JSON that has been output contains: <code>"default_branch": "root"</code>
</li>
<li><code>hub api repos/&lt;org&gt;/&lt;repo&gt; -X GET -F name="&lt;repo&gt;"</code></li>
<li>Note that the JSON that has been output contains: <code>"default_branch": "master"</code>
</li>
</ol>
<p><strong>Via the Web UI</strong></p>
<ol>
<li>Head to the <code>Branches</code> section of your repo settings and update the default branch to <code>root</code>
</li>
<li>Note the confirmation near the top of the page stating that the <code>Default branch changed to root</code>
</li>
<li>Reload the <code>Branches</code> section of the repo settings or head to the <code>Code</code> tab and note that the default branch is still <code>master</code>
</li>
</ol>
<p>I’m mostly wondering if this is unique to our org, or other people are running into this one.</p>
<p>For added context:</p>
<ul>
<li>We do not have any branch protection rules set</li>
<li>We are working with public open-source projects only</li>
<li>We’ve noticed this with normal repos, template repos, and repos created from template repos</li>
</ul>

**Comment by francisfuzz** [comment]
<p>:wave: hello there <a class="mention" href="/gdude2002">@gdude2002</a>, and welcome to the GitHub Support Community! :tada:</p>
<aside class="quote no-group" data-username="gdude2002" data-post="1" data-topic="120244">
<div class="title">
<div class="quote-controls"></div>
 gdude2002:</div>
<blockquote>
<p><strong>Via the Web UI</strong></p>
<ol>
<li>Head to the <code>Branches</code> section of your repo settings and update the default branch to <code>root</code>
</li>
<li>Note the confirmation near the top of the page stating that the <code>Default branch changed to root</code>
</li>
<li>Reload the <code>Branches</code> section of the repo settings or head to the <code>Code</code> tab and note that the default branch is still <code>master</code>
</li>
</ol>
</blockquote>
</aside>
<p>Thanks for sharing those steps here. I attempted to reproduce this as described, but I see that the update propagated. Below is a motion picture of this in action: I’m wondering if you’re still seeing this behavior?</p>
<p></p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/2X/0/0b04e4ec5963365918a1d48b25d3744759af1f95.gif" data-download-href="/uploads/short-url/1ztJKWcsFKUwPINJvIFCPnRp5tj.gif?dl=1" title="web-ui-change-default-branch"><img src="https://user-images.githubusercontent.com/15894826/181084228-b8c8eeeb-61c9-4b6b-ac0d-2660904ff918.gif" alt="web-ui-change-default-branch" data-base62-sha1="1ztJKWcsFKUwPINJvIFCPnRp5tj" width="690" height="383" srcset="https://aws1.discourse-cdn.com/github/optimized/2X/0/0b04e4ec5963365918a1d48b25d3744759af1f95_2_690x383.gif, https://aws1.discourse-cdn.com/github/original/2X/0/0b04e4ec5963365918a1d48b25d3744759af1f95.gif 1.5x, https://aws1.discourse-cdn.com/github/original/2X/0/0b04e4ec5963365918a1d48b25d3744759af1f95.gif 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/2X/0/0b04e4ec5963365918a1d48b25d3744759af1f95_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">web-ui-change-default-branch</span><span class="informations">958×533 1.77 MB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div>
<aside class="quote no-group" data-username="gdude2002" data-post="1" data-topic="120244">
<div class="title">
<div class="quote-controls"></div>
 gdude2002:</div>
<blockquote>
<p><strong>Via the API (using the <code>hub</code> CLI tool)</strong></p>
<ol>
<li><code>hub api repos/&lt;org&gt;/&lt;repo&gt; -X PATCH -F name="&lt;repo&gt;" -F default_branch="root"</code></li>
<li>Note that the JSON that has been output contains: <code>"default_branch": "root"</code>
</li>
<li><code>hub api repos/&lt;org&gt;/&lt;repo&gt; -X GET -F name="&lt;repo&gt;"</code></li>
<li>Note that the JSON that has been output contains: <code>"default_branch": "master"</code>
</li>
</ol>
</blockquote>
</aside>
<p>While I can’t speak to the <code>hub</code> CLI tool, I successfully updated the branch and confirmed it was updated by running the following <code>curl</code> requests:</p>
<pre><code class="lang-auto"># Change the default branch
curl --verbose \
  -H "Authorization: token $PERSONAL_ACCESS_TOKEN" \
  -X PATCH \
  -d '{"default_branch":"root"}' \
  "https://api.github.com/repos/francisfuzz/change-default-branch" 

# Confirm the default branch was changed
curl --verbose \
  -H "Authorization: token $PERSONAL_ACCESS_TOKEN" \
  "https://api.github.com/repos/francisfuzz/change-default-branch"
</code></pre>
<p>I’m wondering if you’re able to reproduce this on a public repository using these <code>curl</code> commands (while changing the appropriate owner and repository name values)? Sharing that would help us investigate further! :male_detective:</p>

**Comment by gdude2002** [comment]
<p>Apologies for not responding for so long - I didn’t get an email for some reason.</p>
<p>I am still noticing this behaviour. An example of a repository I’m seeing it on is <a href="https://github.com/Kotlin-Discord/site" rel="nofollow noopener">this one from the Kotlin-Discord org</a>. This is a public repository, and it’s the repository I tested on in my original post.</p>

**Comment by francisfuzz** [comment]
<aside class="quote no-group" data-username="gdude2002" data-post="3" data-topic="120244">
<div class="title">
<div class="quote-controls"></div>
 gdude2002:</div>
<blockquote>
<p>Apologies for not responding for so long - I didn’t get an email for some reason.</p>
</blockquote>
</aside>
<p>Thanks for following up <a class="mention" href="/gdude2002">@gdude2002</a>! No worries on the timing :v:</p>
<ul>
<li>If you try to change the default branch to <code>root</code> from the <code>Settings</code> tab, what does the UI show? An image or a GIF would help me get a better idea of what’s happening here.</li>
<li>Could you try reproducing this on a new repository and let me know if you’re able to reproduce the same behavior there?</li>
</ul>
<p>Depending on your input, we’ll take the next steps from there :+1:</p>

**Comment by gdude2002** [comment]
<p>Thanks for your quick response!</p>
<p>Of course, Murphy’s law - as soon as I try to record, it works. Genuinely unbelievable, haha.</p>
<p>Do you know if anyone looked into this in the interim and changed anything? I guess it’s going to be difficult to provide extra info when I can no longer reproduce, but at least it works I guess.</p>

**Comment by francisfuzz** [comment, marked as answer]
<aside class="quote no-group" data-username="gdude2002" data-post="5" data-topic="120244">
<div class="title">
<div class="quote-controls"></div>
 gdude2002:</div>
<blockquote>
<p>Of course, Murphy’s law - as soon as I try to record, it works. Genuinely unbelievable, haha.</p>
<p>Do you know if anyone looked into this in the interim and changed anything? I guess it’s going to be difficult to provide extra info when I can no longer reproduce, but at least it works I guess.</p>
</blockquote>
</aside>
<p><a class="mention" href="/gdude2002">@gdude2002</a> - I’m so glad to hear that it works now! :sparkles: While I can’t speak to any one specific change, it’s definitely possible that one of our engineering teams deployed an update between your original post and now. :rocket:</p>
<p>If you’re seeing any other instances of this in the future, please open a new topic with steps to reproduce the behavior and we’ll do our best to help. :v:</p>

**Comment by HazyFish** [comment]
<p></p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/3X/6/0/60a4641a19934f66775443c8d9c4434a496ea509.gif" data-download-href="/uploads/short-url/dMW2Gt33D8miDsdE2swHHnLCS0h.gif?dl=1" title="2ofoW4eEfR" rel="noopener nofollow ugc"><img src="https://user-images.githubusercontent.com/28922700/181084236-a0a0d7af-8135-459f-99a5-614bf20d7444.gif" alt="2ofoW4eEfR" data-base62-sha1="dMW2Gt33D8miDsdE2swHHnLCS0h" width="690" height="370" srcset="https://aws1.discourse-cdn.com/github/optimized/3X/6/0/60a4641a19934f66775443c8d9c4434a496ea509_2_690x370.gif, https://aws1.discourse-cdn.com/github/optimized/3X/6/0/60a4641a19934f66775443c8d9c4434a496ea509_2_1035x555.gif 1.5x, https://aws1.discourse-cdn.com/github/optimized/3X/6/0/60a4641a19934f66775443c8d9c4434a496ea509_2_1380x740.gif 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/3X/6/0/60a4641a19934f66775443c8d9c4434a496ea509_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">2ofoW4eEfR</span><span class="informations">1920×1030 1.4 MB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div><br>
This problem is not fixed!!! I changed the default branch to master but now the default branch is random between the two every time I refresh the page. Same randomness happened on Code and Settings tabs!

**Comment by HazyFish** [comment]
<p></p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/3X/3/7/379580e6ef734b54659dce0a5d5b53a6cf4c491e.gif" data-download-href="/uploads/short-url/7VIznL7bswY49W7lbzGEW4hoq1E.gif?dl=1" title="EdszDaodfq" rel="noopener nofollow ugc"><img src="https://user-images.githubusercontent.com/28922700/181084238-a778e4fd-4c5e-4b7b-95e1-b63281501e6f.gif" alt="EdszDaodfq" data-base62-sha1="7VIznL7bswY49W7lbzGEW4hoq1E" width="690" height="370" srcset="https://aws1.discourse-cdn.com/github/optimized/3X/3/7/379580e6ef734b54659dce0a5d5b53a6cf4c491e_2_690x370.gif, https://aws1.discourse-cdn.com/github/optimized/3X/3/7/379580e6ef734b54659dce0a5d5b53a6cf4c491e_2_1035x555.gif 1.5x, https://aws1.discourse-cdn.com/github/optimized/3X/3/7/379580e6ef734b54659dce0a5d5b53a6cf4c491e_2_1380x740.gif 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/3X/3/7/379580e6ef734b54659dce0a5d5b53a6cf4c491e_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">EdszDaodfq</span><span class="informations">1920×1030 799 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div>

**Comment by gdude2002** [comment]
<p>Just so you’re aware, I worked with GH support via email and the problem turned out to be a Probot config issue — someone had actually configured Probot to reset the branch.</p>
<p>So, if you’re using Probot or some other GH bot/application, it’s always good to check those!</p>

**Comment by ChessDigits** [comment]
<p>I have the same problem, not using Probot. For some reason default branch was develop, trying to change it to master and it just won’t happen.</p>
<p>Note that I imported the repo from a different account, and the visibility was private in the old account and it is public in the current account. Also note that the default branch in the first account was master–I don’t know why it became develop, nor why changing it doesn’t work.</p>

**Comment by nethgato** [comment]
<p>Hi there <a class="mention" href="/chessdigits">@ChessDigits</a>  :wave:</p>
<p>That’s unfortunate to hear that you’re still unable to modify your default branch name! Presumably, you’re updating this via the API. You do also have the option to change via the UI:</p>
<aside class="onebox githubfolder">
  <header class="source">
      <img src="https://docs.github.com/assets/images/site/favicon.svg" class="site-icon" width="16" height="16">
      <a href="https://docs.github.com/en/github/administering-a-repository/changing-the-default-branch" target="_blank" rel="noopener">docs.github.com</a>
  </header>
  <article class="onebox-body">
    

<h3><a href="https://docs.github.com/en/github/administering-a-repository/changing-the-default-branch" target="_blank" rel="noopener">Changing the default branch - GitHub Docs</a></h3>

<p><a href="https://docs.github.com/en/github/administering-a-repository/changing-the-default-branch" target="_blank" rel="noopener">//docs.github.com/en/github/administering-a-repository/changing-the-default-branch</a></p>


  </article>
  <div class="onebox-metadata">
    
    
  </div>
  <div style="clear: both"></div>
</aside>

<p>…but if you’re failing to update via API, perhaps you could provide the output of a <code>curl -v</code> which should reveal what isn’t working properly.</p>
<p>If you do share this output, please be sure to obscure any sensitive data (like a PAT that might be used for API access) when you post.</p>
<p>Cheers!</p>

**Comment by ChessDigits** [comment]
<p>I don’t know what you mean with the API. I’m changing it directly from the GitHub website­ (repo &gt; settings &gt; branches). Now that being said it does look like it caught on, thought it does happen sometimes it’ll still show develop… Seriously there seems to be a bug with this feature. Anyway, no need to reply anymore, you do what you want with the feedback.</p>

**Comment by nethgato** [comment]
<p>Hi <a class="mention" href="/chessdigits">@ChessDigits</a> !</p>
<p><s>I mention the API, because the OP for this thread is about updating the default branch specifically via our API and not via the UI.</s></p>
<p>Edit: the primary focus of this thread is about behavior relating to the API, and outdated experiences from the UI that should no longer occur.</p>
<p>Perhaps we should get a new thread started in the GitHub Help <a href="https://github.community/c/github-help/how-to-use-git-and-github/21">How to Use Git and GitHub</a> board.</p>
<p>Though I’m also happy to accept your feedback, but I’m not quite sure what’s failing so far. Would you mind sharing screenshots of the flow in the UI? If any errors are elevated, seeing those errors in the dev console from the browser would be interesting!</p>
<p>As an aside, I went back over my personal repositories that were still named <code>master</code> and I updated them to <code>main</code> via the UI without any issue. <a href="https://github.com/nethgato/PublicTest">Here’s an example I modified earlier</a>, before I replied to your post.</p>

**Comment by xero** [comment]
necro bumping this for anyone who google's there way here like i did. 

the issue is the two `-F` flags. the repo name is implied from the third positional argument. 
simply pass `default_branch` and it works as expected.

`gh api repos/<org>/<repo> -X PATCH -F default_branch="root`

**Reply by medley56** [reply]
This worked for me. The web UI was definitely broken for me.

**Comment by scoful** [comment]
As of now (20250227), the correct location is: Settings - General - Default branch (it's under the General category, not under Branches).
I hope this helps others who encounter the same issue and saves them time.

**Reply by owenfar** [reply]
Even their own docs are wrong: https://docs.github.com/en/repositories/configuring-branches-and-merges-in-your-repository/managing-branches-in-your-repository/changing-the-default-branch

Thanks for the heads-up, was lucky to quickly find this answer

**Comment by medley56** [comment]
Just commenting to confirm that this is still an issue in the Web UI. The API call mentioned by @xero works.
